### DISCUSSION TITLE
How to couple an outer existing code into MOOSE framework

### DISCUSSION THREAD

**Discussion by LagrangeW** [idaholab/moose#16438] [Q&A General] [answered]
Dear MOOSE Experts,

I've read some articles about coupling an existing code into MOOSE framework , but they all mention less about the process of  the wrapping of the outer code . And I can't find any useful Doc in the MOOSE page.

So would you please tell me some useful docs or pages I can refer to about how to couple an outer code into MOOSE framework?

Thank you so much!

Wayne Wang

**Comment by aeslaughter** [comment, marked as answer]
Here is our documentation: https://mooseframework.inl.gov/application_development/moose_wrapped_apps.html

**Reply by LagrangeW** [reply]
Thanks so much for your patience! I've seen this before, but I think It's just not that detailed， it just give some information about wrapping interface ,but not mention the detail about how the external code should be departed or something else...It's just my thought. Thanks for your patience again!

**Reply by aeslaughter** [reply]
Unfortunately, that is the best we have at the moment. It is very difficult to write documentation on wrapping applications. The structure of the application being wrapped as well as the problem(s) to be solved are completely unknown. If you have a specific application in mind and want some guidance I suggest starting a new Discussion. If you can't share then please reach out to us via email.

**Reply by LagrangeW** [reply]
Got it, thank you for your support! I'll learn more about it

**Reply by fdkong** [reply]
module "`petsc_external_solver`" is the simplest example we have.  There are already several successful applications that followed the example.

**Reply by LagrangeW** [reply]
Thank you so much ! But actually I don't know where to find it ,  will you please to be more specific?  I'd appreciate i t!

**Reply by fdkong** [reply]
https://github.com/idaholab/moose/tree/next/modules/external_petsc_solver

**Reply by LagrangeW** [reply]
Got it ! Thank you so much !
